src/Razor/src/Microsoft.AspNetCore.Razor.LanguageServer/Refactoring/ExtractToCodeBehindCodeActionResolver.cs

```diff
@@ -0,0 +1,168 @@
+            }
+
+            _foregroundDispatcher = foregroundDispatcher;
+            _documentResolver = documentResolver;
+        }
+
+        public override async Task<WorkspaceEdit> ResolveAsync(JObject data, CancellationToken cancellationToken)
+        {
+            var actionParams = data.ToObject<ExtractToCodeBehindParams>();
+            var path = Path.GetFullPath(actionParams.Uri.LocalPath);
+
+            var document = await Task.Factory.StartNew(() =>
+            {
+                _documentResolver.TryResolveDocument(path, out var documentSnapshot);
+                return documentSnapshot;
+            }, cancellationToken, TaskCreationOptions.None, _foregroundDispatcher.ForegroundScheduler);
+            if (document is null)
+            {
+                return null;
+            }
+
+            var text = await document.GetTextAsync();
+            if (text is null)
+            {
+                return null;
+            }
+
+            var codeDocument = await document.GetGeneratedOutputAsync();
+            if (codeDocument.IsUnsupported())
+            {
+                return null;
+            }
+
+            if (!FileKinds.IsComponent(codeDocument.GetFileKind()))
+            {
+                return null;
+            }
+
+            var codeBehindPath = "";
+            var n = 0;
+            do
+            {
+                var identifier = n > 0 ? n.ToString() : "";  // Make it look nice
+                codeBehindPath = Path.Combine(
+                    Path.GetDirectoryName(path),
+                    $"{Path.GetFileNameWithoutExtension(path)}{identifier}{Path.GetExtension(path)}.cs");
+                n++;
+            } while (File.Exists(codeBehindPath));
+
+            var codeBehindUri = new UriBuilder()
+            {
+                Scheme = Uri.UriSchemeFile,
+                Path = codeBehindPath,
+                Host = string.Empty,
+            }.Uri;
+
+            var className = Path.GetFileNameWithoutExtension(path);
+            var contents = text.GetSubTextString(new CodeAnalysis.Text.TextSpan(actionParams.ExtractStart, actionParams.ExtractEnd - actionParams.ExtractStart));
+            var compilationUnit = GenerateCodeBehindClass(className, contents, codeDocument);
+
+            var changes = new Dictionary<Uri, IEnumerable<TextEdit>>
+            {
+                [actionParams.Uri] = new[]
+                {
+                    new TextEdit()
+                    {
+                        NewText = "",
+                        Range = codeDocument.RangeFromIndices(actionParams.RemoveStart, actionParams.RemoveEnd)
+                    }
+                },
+                [codeBehindUri] = new[]
+                {
+                    new TextEdit()
+                    {
+                        NewText = compilationUnit.NormalizeWhitespace().ToFullString(),
+                        Range = new Range(new Position(0, 0), new Position(0, 0))
```

Comment:
Up at the top you can do a `private static readonly Range StartOfDocumentRange = new Range(new Position(0, 0), new Position(0, 0));`

so you don't need to recompute this and it's extra clear